tests/ops/op_math/test_controlled.py

```diff
@@ -1608,146 +1913,288 @@ def test_qubit_unitary(M):
+            for op_ in qml.CRot.compute_decomposition(0.1, 0.2, 0.3, wires=[7, 0]):
+                qml.ctrl(op_, control=3)
+        tape_expected = QuantumScript.from_queue(q_tape)
 
-    with qml.queuing.AnnotatedQueue() as q_tape:
-        ctrl(qml.CY, 0)(wires=[1, 2])
+        def stopping_condition(o):
+            return not isinstance(o, Controlled) or not o.has_decomposition
 
-    tape = QuantumScript.from_queue(q_tape)
-    tape = tape.expand()
+        actual = tape.expand(depth=10, stop_at=stopping_condition)
+        expected = tape_expected.expand(depth=10, stop_at=stopping_condition)
+        actual_mat = qml.matrix(actual, wire_order=[3, 7, 0])
+        expected_mat = qml.matrix(expected, wire_order=[3, 7, 0])
+        assert qml.math.allclose(actual_mat, expected_mat, atol=tol, rtol=0)
 
-    assert len(tape.operations) == 2
+    @pytest.mark.parametrize(
+        "op",
+        [
+            qml.ctrl(qml.ctrl(qml.S, 7), 3),  # nested control
+            qml.ctrl(qml.S, [3, 7]),  # multi-wire control
+        ],
+    )
+    def test_nested_ctrl_containing_phase_shift(self, op):
+        """Test that nested controlled ops are expanded correctly when phase shift is involved
 
-    assert tape.operations[0].name == "C(CRY)"
-    assert tape.operations[1].name == "C(S)"
+        The decomposition of S gate contains a PhaseShift. In the nested case, we do not want
+        to apply control to the expanded PhaseShift like how it is typically done for other
+        operations, because the decomposition of PhaseShift contains a GlobalPhase, the controlled
+        version of which we do not have handling for.
 
+        TODO: remove this special case once ControlledGlobalPhase is implemented.
 
-def test_ctrl_template():
-    """Test that a controlled template correctly expands
-    on a device that doesn't support it"""
+        """
 
-    weights = np.ones([3, 2])
+        with qml.queuing.AnnotatedQueue() as q_tape:
+            op(wires=0)
 
-    with qml.queuing.AnnotatedQueue() as q_tape:
-        ctrl(qml.templates.BasicEntanglerLayers, 0)(weights, wires=[1, 2])
+        tape = QuantumScript.from_queue(q_tape)
+        assert tape.expand(depth=1).circuit == [
+            Controlled(qml.ControlledPhaseShift(np.pi / 2, wires=[7, 0]), control_wires=[3])
+        ]
 
-    tape = QuantumScript.from_queue(q_tape)
-    tape = expand_tape(tape, depth=2)
-    assert len(tape) == 9
-    assert all(o.name in {"CRX", "Toffoli"} for o in tape.operations)
+        assert tape.expand(depth=2).circuit == [
+            qml.ControlledPhaseShift(np.pi / 4, wires=[3, 7]),
+            qml.Toffoli(wires=[3, 7, 0]),
+            qml.ControlledPhaseShift(-np.pi / 4, wires=[3, 0]),
+            qml.Toffoli(wires=[3, 7, 0]),
+            qml.ControlledPhaseShift(np.pi / 4, wires=[3, 0]),
+        ]
 
+    def test_adjoint_of_ctrl(self):
+        """Tests that adjoint(ctrl(fn)) and ctrl(adjoint(fn)) are equivalent"""
```

Comment:
Here we test their equivalence only in conjunction with `qml.simplify`, would it be reasonable to make them equivalent without a call to simplify?

Reply:
They are actually not equivalent without the call to `qml.simplify`. I mean they are equivalent as in they have the same effect, but they wouldn't be the exact same two sequences of operations. This might be something worth investigating but it would be outside the scope of this PR. This particular test case is not new, it was copied over from another place.

Reply:
Yep agreed, it probably is it out of scope for this PR. I just brought it up here when I saw this test because it was something I also suggested to you in a our private conversation, but that could be of interest to other people.

Specifically, arbitrary chains of `C(A(C(C(C(A(...))))))` could be simplified automatically upon construction if we defined a **canonical ordering** for `Adjoint` and `Controlled`. Any such chain could then be simplified to at-most 1 adjoint and 1 control operator. For the given example, if the canonical order is for adjoint to be inside, we could automatically have the following steps take place:
```
-> C(A(C(A(...)))) (after the existing simplification of nested control)
-> C(C(A(A(...)))) 
-> C(C(A(...))) (if adjoint simplification is implemented?)
-> C(A(...))
```
Whereas with this PR we are stuck at the first step.

Reply:
This looks nice, but at the same time if some users/developers still use the `Adjoint` constructor, then we won't be able to assume this canonical ordering everywhere unless we enforce this ordering in the `Adjoint` constructor instead.

Reply:
So `qml.simplify` and `Adjoint.decomposition` would still turn `Adjoint(Controlled)` into `Controlled(Adjoint)` simply by updating the adjoint method.  So we would still be able to handle `Adjoint(Controlled)` just fine. The idea is "when" do we handle this? Lazily, or when requested in a decomposition?

Reply:
I think similar to what I did in this PR, I would hope that `Adjoint(Controlled)` and `Controlled(Adjoint)` would decompose into the same thing, instead of having `Adjoint(Controlled)` decompose into `Controlled(Adjoint)`, unless I misunderstood what you mean.